File: mobile/lib/features/fuel/voice/parsed_fuel_transaction_dto.dart
/// Mirror BE `ParsedFuelTransactionDto`. `amountVnd == null` → parse fail field bắt buộc,
/// mobile show fail dialog. `error != null` cho biết lý do (whisper_*, audio_*, transcribe_empty).
class ParsedFuelTransactionDto {
  const ParsedFuelTransactionDto({
    required this.rawText,
    required this.transactionDate,
    this.amountVnd,
    this.odometerKm,
    this.missingRequiredFields = const [],
    this.error,
  });

  final String rawText;
  final int? amountVnd;
  final int? odometerKm;
  final DateTime transactionDate;
  final List<String> missingRequiredFields;
  final String? error;

  bool get hasAmount => amountVnd != null && amountVnd! > 0;

  factory ParsedFuelTransactionDto.fromJson(Map<String, dynamic> json) {
    return ParsedFuelTransactionDto(
      rawText: (json['rawText'] as String?) ?? '',
      amountVnd: (json['amountVnd'] as num?)?.toInt(),
      odometerKm: (json['odometerKm'] as num?)?.toInt(),
      transactionDate: DateTime.tryParse(json['transactionDate'] as String? ?? '') ?? DateTime.now(),
      missingRequiredFields: ((json['missingRequiredFields'] as List?) ?? const [])
          .whereType<String>()
          .toList(),
      error: json['error'] as String?,
    );
  }
}

File: backend/src/Httm.XangDau.Api/Features/Fuel/Voice/Contracts/ParsedFuelTransactionDto.cs
namespace Httm.XangDau.Api.Features.Fuel.Voice.Contracts;

/// <summary>
/// Response của <c>POST /api/fuel/voice/parse-fuel-tx</c>. Field nullable nếu không
/// nghe được — mobile nhận DTO này, prefill form, user xem rồi bấm Lưu.
/// </summary>
public sealed class ParsedFuelTransactionDto
{
    /// <summary>Text Whisper nghe được — luôn có để show user khi parse fail.</summary>
    public required string RawText { get; init; }

    /// <summary>Số tiền (VNĐ) — bắt buộc cho form. Null = parse fail → mobile show fail dialog.</summary>
    public long? AmountVnd { get; init; }

    /// <summary>Số công tơ km — optional.</summary>
    public long? OdometerKm { get; init; }

    /// <summary>Ngày giao dịch — default = today nếu không nói tới ngày.</summary>
    public DateTime TransactionDate { get; init; }

    /// <summary>Liệt kê các field bắt buộc bị missing — mobile dùng để hiển dialog phù hợp.</summary>
    public IReadOnlyList<string> MissingRequiredFields { get; init; } = [];

    /// <summary>Mã lỗi nội bộ (whisper_unreachable, audio_too_large, ...) — null nghĩa là OK.</summary>
    public string? Error { get; init; }
}
